src/backend/utils/adt/datetime.c

```diff
@@ -1963,6 +1967,16 @@ DecodeTimeOnly(char **field, int *ftype, int nf,
 				if (tzp == NULL)
 					return DTERR_BAD_FORMAT;
 
+				/*
+				 * For date format like 'yyyy-mm-dd', we should return time 00:00:00.0000000
+				 * instead of marking it as invalid time format
+				 */
```

Comment:
I think you might need to add another format check here since if the cast string comes from a variable, it will pass the antlr check and only reveal its value during execution.